tests/posix/env/src/env.c

```diff
@@ -0,0 +1,182 @@
+/*
+ * Copyright (c) 2023, Meta
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <stdio.h>
+#include <stdlib.h>
+
+#include <zephyr/ztest.h>
+
+#define _m_home "/home/zephyr"
+#define _m_uid  STRINGIFY(1000)
+#define _m_pwd  "/tmp"
+
+#define _m_alt_home "/this/path/is/much/longer/than" _m_home
+
+#define DEFINE_ENVIRON(_handle, _key, _val) char _handle[] = STRINGIFY(_key) "=" _val
+#define RESET_ENVIRON(_handle, _key, _val)                                                         \
+	snprintf(_handle, ARRAY_SIZE(_handle), "%s=%s", _key, _val)
+
+#if defined(CONFIG_NEWLIB_LIBC) || defined(CONFIG_PICOLIBC)
+/* newlib headers seem to be missing this */
+int getenv_r(const char *name, char *val, size_t len);
+#endif
+
+extern char **environ;
+static char **old_environ;
+
+static DEFINE_ENVIRON(home, "HOME", _m_home);
+static DEFINE_ENVIRON(uid, "UID", _m_uid);
+static DEFINE_ENVIRON(pwd, "PWD", _m_pwd);
+
+static char *environ_for_test[] = {home, uid, pwd, NULL};
+
+ZTEST(env, test_getenv)
+{
+	zassert_equal(getenv(NULL), NULL);
+	zassert_equal(getenv(""), NULL);
+	zassert_equal(getenv("invalid=key"), NULL);
+	zassert_equal(getenv("HOME=" _m_home), NULL);
+	zassert_equal(getenv("PWDR"), NULL);
+
+	zassert_mem_equal(getenv("HOME"), _m_home, strlen(_m_home) + 1);
+	zassert_mem_equal(getenv("UID"), _m_uid, strlen(_m_uid) + 1);
+	zassert_mem_equal(getenv("PWD"), _m_pwd, strlen(_m_pwd) + 1);
+}
+
+ZTEST(env, test_getenv_r)
+{
+	static char buf[16];
+	static const int exp_errno[] = {
+		EINVAL, EINVAL, EINVAL, EINVAL, ENOENT, ENOENT, ENOENT, EINVAL, EINVAL, EINVAL,
+	};
+	static const struct args_s {
+		const char *name;
+		char *buf;
+		size_t size;
+	} args[] = {
+		/* invalid input */
+		{NULL, NULL, 0},
+		{NULL, NULL, 42},
+		{NULL, buf, 0},
+		{NULL, buf, sizeof(buf)},
+		{"hello", NULL, 0},
+		{"hello", NULL, 42},
+		{"hello", buf, 0},
+
+		/* invalid names */
+		{"", buf, sizeof(buf)},
+		{"invalid=key", buf, sizeof(buf)},
+		{"HOME=", buf, sizeof(buf)},
+	};
+
+	BUILD_ASSERT(ARRAY_SIZE(exp_errno) == ARRAY_SIZE(args));
+
+	ARRAY_FOR_EACH(args, i)
```

Comment:
Nit: Using macros to replace trivial constructs is bad for readability.
Everybody who knows C can understand at a glance what
for (int i = 0; i < ARRAY_SIZE(args) ; i++) 
does.
But only those who know this macro know what it hides.

Reply:
I'm perfectly fine using this macro in code that I maintain.

I don't really consider this an issue.